ERT-904: Local config work

arielalmendral opened the pull request:
1. Possible to attach any obsset to any ministep
2. Ert script support to basic local config functionalities.

Please comment, thanks!

joakim-hove commented:
Sorry about the slow response - will look properly on this tomorrow latest.

joakim-hove commented:
Please apply this: https://github.com/joakim-hove/ert/commit/ef019c4369a0482ad28d97017ecd8759b304ecb6

joakim-hove commented on devel/python/python/ert/enkf/local_dataset.py:
```diff
@@ -0,0 +1,57 @@
+
+class LocalDataset(BaseCClass):
+
+    def __init__(self, name):
```
Please implement the `__contains__` method (requires a C function as well), so that we can check if a certain key is valid.

joakim-hove commented on devel/python/python/ert/enkf/local_dataset.py:
```diff
@@ -0,0 +1,57 @@
+    def getName(self):
+        return LocalDataset.cNamespace().name(self)
+    
+    def getActiveList(self, key):
```
Please add some safety here:
```python
if key in self:
    return LocalDataset.cNamespace().active_list(self , key)
else:
    raise KeyError("Local key:%s not recognized" % key)
```

joakim-hove commented on devel/python/python/ert/enkf/local_dataset.py:
```diff
@@ -0,0 +1,57 @@
+
+
+cwrapper = CWrapper(ENKF_LIB)
+cwrapper.registerType("local_dataset", LocalDataset)
```
Use the method `registerObjectType` which is a convenience wrapper registering all three types in one go.

joakim-hove commented on devel/python/python/ert/enkf/local_ministep.py:
```diff
@@ -0,0 +1,44 @@
+        LocalMinistep.cNamespace().free(self) 
+
+cwrapper = CWrapper(ENKF_LIB)
+cwrapper.registerType("local_ministep", LocalMinistep)
```
`registerObjectType`

joakim-hove commented on devel/python/python/ert/enkf/local_updatestep.py:
```diff
@@ -0,0 +1,31 @@
+        LocalUpdateStep.cNamespace().free(self) 
+
+cwrapper = CWrapper(ENKF_LIB)
+cwrapper.registerType("local_updatestep", LocalUpdateStep)
```
`registerObjectType`

joakim-hove commented on devel/python/tests/ert/enkf/test_local_config.py:
```diff
@@ -0,0 +1,56 @@
+
+
+    def test_local_config( self ):
+        with TestAreaContext("enkf_test", store_area=True) as work_area:
```
Here you can alternatively use `ErtTestContext` - then you do not need to manually copy the directory, and the main instance is automagically instantiated:
```
with ErtTestContext("manual_load_test", self.config_file, store_area = True) as test_context:
   main = test_context.getErt()
   ....
```
If using `ErtTestContext` the whole rest of the test should be in the `with ... :` scope.

I guess the `store_area = True` is leftover from debugging? Should be set to `False`


```

joakim-hove commented on devel/python/tests/ert/enkf/test_local_config.py:
```diff
@@ -0,0 +1,56 @@
+        sfile = self.createTestPath("test_local_config.txt")                        
+        local_config.writeLocalConfigFile(sfile)            
+        self.assertTrue( os.path.exists( sfile ))                     
+        main.free()
```
No explicit free in top level code!

joakim-hove commented:
I think this looks very good - and apart from the minor comments mentioned I am ready to merge this.

One question - just for my own understanding: The solution chosen here is to let the `LocalConfig`object instantiate all child objects - and then consequently let that top level C structure do all life time management? It is certainly the simplest way to get it Right &trade;.

joakim-hove commented on devel/python/python/ert/enkf/local_config.py:
```diff
@@ -14,36 +14,95 @@
+        assert isinstance(filename, str)
+        LocalConfig.cNamespace().write_local_config_file(self, filename)
+           
+    def createUpdatestep(self, update_step_key):
```
I guess it would be good to have `getUpdateStep(), getMinistep(), getObsdata()` and `getDataset()`methods.

arielalmendral commented:
Thanks for the comments Joakim! Yes, LocalConfig is instantianing the objects for simplicity. I will add/fix according to your comments and let you know.

arielalmendral commented:
Hi!

I have now addressed your comments in a new revision so that you can compare. Let me know if you prefer me to squash these two revisions (the one you commented and the new one) into one.

Also, some more work is done on test_local_config.py using a synthetic data set available under 'test-data/local/custom_kw/'. Note that this test is ongoing work.

joakim-hove commented:
Cool - I am ready to merge this now; no squash required.

>Also, some more work is done on test_local_config.py using a synthetic data set available under 'test-data/local/custom_kw/'. Note that this test is ongoing work.

OK - should I wait for that - or just merge now?

arielalmendral commented:
Please merge now :-)
